### PR TITLE
Make sure parallel mode can be used with sphinx-automodapi

### PR DESCRIPTION
**Pull request by astrofrog**
Currently sphinx can't run in parallel mode when using sphinx-automodapi because extensions have to declare they are parallel-safe. However, I discovered that ``smart_resolver`` is not parallel-safe since it relies on populating a global dictionary which doesn't get communicated back to the main process. Instead I've implemented a file-based approach for the dictionaries, with one file per process.

For the astropy docs, using parallel mode ``-j4`` on a 4-core mac laptop I get a speedup of 40%:

```
real	4m45.616s
user	12m12.225s
sys	0m50.415s
```

compared to the serial mode:

```
real	6m37.317s
user	5m51.115s
sys	0m28.985s
```

I need to investigate why the speedup isn't more significant.

### CODE REVIEW

**Comment by astrofrog** [comment]
@eteq - since you have done a lot with the sphinx extensions in the past, could you review this?

File: sphinx_automodapi/tests/test_cases.py [hunk 1]
```diff
@@ -110,6 +117,8 @@ def test_run_full_case(tmpdir, case_dir):
             path_relative = os.path.relpath(path_reference, output_dir)
             path_actual = os.path.join(docs_dir, path_relative)
             assert os.path.exists(path_actual)
-            actual = io.open(path_actual, encoding='utf8').read()
-            reference = io.open(path_reference, encoding='utf8').read()
+            with io.open(path_actual, encoding='utf8') as f:
```
Comment (eteq) [inline sphinx_automodapi/tests/test_cases.py]: I could be wrong here, but I think this *might* cause trouble in locales where UTF8 doesn't come out as a default?  Might it be sufficient to just do a Byte-wise comparison to sidestep the encoding problem?

File: sphinx_automodapi/smart_resolver.py [hunk 2]
```diff
@@ -13,22 +13,58 @@
 corrected if possible.
 """
 
+import os
+import glob
+import uuid
+import tempfile
 from docutils.nodes import literal, reference
 
+# NOTE: This extension originally used a dictionary stored in app.env to keep
```
Comment (eteq) [inline sphinx_automodapi/smart_resolver.py]: This whole approach seems a bit problematic, as it depends on file systems and sort of implementation details.  Is there a reason you didn't try using the`event-env-merge-info` event to do the parallel gathering (http://www.sphinx-doc.org/en/stable/extdev/appapi.html#event-env-merge-info)?  I thought it was meant to solve exactly this problem? (Although I've not used it myself...)

**Comment by astrofrog** [comment]
This was closed as the master branch was updated to include history. I'll need to do a new implementation based on @eteq's comments anyway.

**Pull request closed without merging**